### PR TITLE
Bail out jump threading on indirect branches

### LINKED ISSUE #102351: Miscompile in code with indirect gotos
The simplifycfg pass (incorrectly?) optimizes one of the case statements.

```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@.str = private unnamed_addr constant [23 x i8] c"OP_1:(instruction=%d)\0A\00", align 1
@.str.1 = private unnamed_addr constant [28 x i8] c"TERMINATE:(instruction=%d)\0A\00", align 1

; Function Attrs: mustprogress norecurse uwtable
define dso_local noundef i32 @main() #0 {
entry:
  %bytecode = alloca [2 x ptr], align 16
  store ptr blockaddress(@main, %VM__OP_1), ptr %bytecode, align 16, !tbaa !5
  %arrayidx1 = getelementptr inbounds [2 x ptr], ptr %bytecode, i64 0, i64 1
  store ptr blockaddress(@main, %VM__TERMINATE), ptr %arrayidx1, align 8, !tbaa !5
  br label %while.body

while.body:                                       ; preds = %entry, %sw.epilog
  %state.0 = phi i32 [ 0, %entry ], [ %state.1, %sw.epilog ]
  %index.0 = phi i32 [ 0, %entry ], [ %index.2, %sw.epilog ]
  switch i32 %state.0, label %sw.epilog [
    i32 0, label %sw.bb
    i32 1, label %VM__OP_1
    i32 2, label %sw.bb4
  ]

sw.bb:                                            ; preds = %while.body
  br label %indirectgoto

VM__OP_1:                                         ; preds = %while.body, %indirectgoto
  %index.1 = phi i32 [ %index.3, %indirectgoto ], [ %index.0, %while.body ]
  br label %sw.epilog

sw.bb4:                                           ; preds = %while.body
  %call = call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef %index.0)
  %inc = add nsw i32 %index.0, 1
  br label %indirectgoto

sw.epilog:                                        ; preds = %while.body, %VM__OP_1
  %state.1 = phi i32 [ %state.0, %while.body ], [ 2, %VM__OP_1 ]
  %index.2 = phi i32 [ %index.0, %while.body ], [ %index.1, %VM__OP_1 ]
  br label %while.body, !llvm.loop !9

VM__TERMINATE:                                    ; preds = %indirectgoto
  %call7 = call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str.1, i32 noundef %index.3)
  ret i32 0

indirectgoto:                                     ; preds = %sw.bb4, %sw.bb
  %index.3 = phi i32 [ %inc, %sw.bb4 ], [ %index.0, %sw.bb ]
  %idxprom.pn = sext i32 %index.3 to i64
  %indirect.goto.dest.in = getelementptr inbounds [2 x ptr], ptr %bytecode, i64 0, i64 %idxprom.pn
  %indirect.goto.dest = load ptr, ptr %indirect.goto.dest.in, align 8, !tbaa !5
  indirectbr ptr %indirect.goto.dest, [label %VM__OP_1, label %VM__TERMINATE]
}
declare i32 @printf(ptr noundef, ...) #1

attributes #0 = { mustprogress norecurse uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 8, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 18.0.0git (https://github.com/llvm/llvm-project.git 67782d2de5ea9c8653b8f0110237a3c355291c0e)"}
!5 = !{!6, !6, i64 0}
!6 = !{!"any pointer", !7, i64 0}
!7 = !{!"omnipotent char", !8, i64 0}
!8 = !{!"Simple C++ TBAA"}
!9 = !{!10, !10, i64 0}
!10 = !{!"int", !7, i64 0}
!11 = distinct !{!11, !12, !13}
!12 = !{!"llvm.loop.mustprogress"}
!13 = !{!"llvm.loop.unroll.disable"}
```

$ opt -passes=simplifycfg orig.ll -S -o out.ll
```llvm
; ModuleID = 'orig.ll'
source_filename = "orig.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@.str = private unnamed_addr constant [23 x i8] c"OP_1:(instruction=%d)\0A\00", align 1
@.str.1 = private unnamed_addr constant [28 x i8] c"TERMINATE:(instruction=%d)\0A\00", align 1

; Function Attrs: mustprogress norecurse uwtable
define dso_local noundef i32 @main() #0 {
entry:
  %bytecode = alloca [2 x ptr], align 16
  store ptr blockaddress(@main, %VM__OP_1), ptr %bytecode, align 16, !tbaa !5
  %arrayidx1 = getelementptr inbounds [2 x ptr], ptr %bytecode, i64 0, i64 1
  store ptr blockaddress(@main, %VM__TERMINATE), ptr %arrayidx1, align 8, !tbaa !5
  br label %while.body

while.body:                                       ; preds = %sw.epilog, %entry
  %state.0 = phi i32 [ 0, %entry ], [ %state.1, %sw.epilog ]
  %index.0 = phi i32 [ 0, %entry ], [ %index.0, %sw.epilog ]
  switch i32 %state.0, label %sw.epilog [
    i32 0, label %VM__TERMINATE
    i32 1, label %VM__OP_1
    i32 2, label %sw.bb4
  ]

VM__OP_1:                                         ; preds = %while.body
  br label %sw.epilog

sw.bb4:                                           ; preds = %while.body
  %call = call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef %index.0)
  %inc = add nsw i32 %index.0, 1
  br label %VM__TERMINATE

sw.epilog:                                        ; preds = %VM__OP_1, %while.body
  %state.1 = phi i32 [ %state.0, %while.body ], [ 2, %VM__OP_1 ]
  br label %while.body, !llvm.loop !9

VM__TERMINATE:                                    ; preds = %sw.bb4, %while.body
  %index.3 = phi i32 [ %inc, %sw.bb4 ], [ %index.0, %while.body ]
  %call7 = call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str.1, i32 noundef %index.3)
  ret i32 0
}

declare i32 @printf(ptr noundef, ...) #1

attributes #0 = { mustprogress norecurse uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 8, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 18.0.0git (https://github.com/llvm/llvm-project.git 67782d2de5ea9c8653b8f0110237a3c355291c0e)"}
!5 = !{!6, !6, i64 0}
!6 = !{!"any pointer", !7, i64 0}
!7 = !{!"omnipotent char", !8, i64 0}
!8 = !{!"Simple C++ TBAA"}
!9 = !{!10, !10, i64 0}
!10 = !{!"int", !7, i64 0}
```

### PR DESCRIPTION
**Pull request by hiraditya**
The bug was introduced by https://github.com/llvm/llvm-project/pull/68473

Fixes: #102351

### CODE REVIEW

File: llvm/test/Transforms/SimplifyCFG/switch-branch-fold-indirectbr-102351.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/SimplifyCFG/switch-branch-fold-indirectbr-102351.ll:8]: Please remove unnecessary attributes/metadata.

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/Local.cpp:1037]: Can you explain this change?

Comment (nikic) [inline llvm/lib/Transforms/Utils/Local.cpp:1037]: This change makes no sense -- did you run the tests for this PR at all?

This is finding a single common predecessor, so it bails out if there are multiple. After your change it bails out if there are any common predecessors.

**Review by nikic — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (hiraditya) [reply llvm/lib/Transforms/Utils/Local.cpp:1037]: yeah, this is not needed. The second bug is somewhere else in simplifycfg, i'm still investigating.

**Comment by dtcxzyw** [comment]
> Failed Tests (1):
  LLVM :: Transforms/SimplifyCFG/switch-branch-fold-indirectbr-102351.ll

**Comment by hiraditya** [comment]
> > Failed Tests (1):
> > LLVM :: Transforms/SimplifyCFG/switch-branch-fold-indirectbr-102351.ll

yeah, another bug that i still need to find.

**Comment by dtcxzyw** [comment]
> Failed Tests (3):
  LLVM :: Transforms/JumpThreading/pr40992-indirectbr-folding.ll
  LLVM :: Transforms/SimplifyCFG/indirectbr.ll
  LLVM :: Transforms/SimplifyCFG/switch-branch-fold-indirectbr-102351.ll

**Comment by nikic** [comment]
Please rebase this to current main (it looks like CI is broken on your base revision), minimize the test and use UTC.

**Comment by hiraditya** [comment]
> Please rebase this to current main (it looks like CI is broken on your base revision), minimize the test and use update_test_checks.

I minimized a few instructions, lmk if you need it to be smaller

**Comment by hiraditya** [comment]
Merging this assuming there are no further comments for changes. Please let me know any more changes are required and i'll follow up.

**Comment by SanjaLV** [comment]
Hey @tru is there any chance this fix goes into LLVM stable release?

This is regression from the LLVM 17 that miscompiles a lot of non-trivial indirect goto code.

**Comment by tru** [comment]
Hi, since we are wrapping up this next release we are very strict with the fixes we pick at this point. Can you please respond to the following questions to help me understand if this has to be included in the final release or not.

Is this PR a fix for a regression or a critical issue?

What is the risk of accepting this into the release branch?

What is the risk of NOT accepting this into the release branch?

**Comment by hiraditya** [comment]
> This is likely to be reverted due to code review policy violation

It didn't occur to me that there would be any violation of some sorts. I pushed the latest change last week and there wasn't any feedback since then.

**Comment by aemerson** [comment]
ping @nikic @hiraditya what's going on with this then?

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/Local.cpp:1032]: I don't understand why phi blocks in the predecessor are relevant here. Your test case happens to have a phi, but if I replace the phi with an argument I still get the same miscompile after your patch.

**Comment by nikic** [comment]
Sorry for not following up here before...

> It didn't occur to me that there would be any violation of some sorts. I pushed the latest change last week and there wasn't any feedback since then.

Please see https://llvm.org/docs/CodeReview.html#code-review-workflow. If you submit a PR, you need to wait for an explicit approval. Ping the PR if needed. Lack of response is not the same as approval.

This PR did not even take into account the feedback that *was* provided. If the test case had been properly minimized, it would have been clearer that this is not the correct bailout condition.

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (hiraditya) [reply llvm/lib/Transforms/Utils/Local.cpp:1032]: do you mind sharing your test case? I tried replacing switch with if-else and moving the check to a function parameter but didn't get a miscompile. https://godbolt.org/z/xvs571nbj

```cpp
#include<stdio.h>
__attribute__((noinline))
int foo(int state) {
  void* bytecode[2];
  bytecode[0] = &&VM__OP_1;
  bytecode[1] = &&VM__TERMINATE;

  //int state = 0;
  int index = 0;

  while (1) {
    if (state==0) {
      goto *bytecode[index];
    }
    if (state == 1) {
      printf("Unreachable\n");
      // NOTE: THIS IS ONLY REACHABLE VIA INDIRECT GOTOS
    VM__OP_1:
      printf("VM__OP_1\n");
      state = 2;
    }
    else if (state== 2) {
      printf("OP_1:(instruction=%d)\n", index);
      index++;
      goto *bytecode[index];
    }
  }

VM__TERMINATE:
  printf("TERMINATE:(instruction=%d)\n", index);
  return 0;
}

int main() {
  foo(0);
  return 0;
}```

**Comment by hiraditya** [comment]
> Sorry for not following up here before...
> 
> > It didn't occur to me that there would be any violation of some sorts. I pushed the latest change last week and there wasn't any feedback since then.
> 
> Please see https://llvm.org/docs/CodeReview.html#code-review-workflow. If you submit a PR, you need to wait for an explicit approval. Ping the PR if needed. Lack of response is not the same as approval.

Ack.

> This PR did not even take into account the feedback that _was_ provided. If the test case had been properly minimized, it would have been clearer that this is not the correct bailout condition.

My bad, i'll try reducing the test case further. First attempt at reducing it was unsuccessful, commented as a follow up in your inline comment.

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (hiraditya) [reply llvm/lib/Transforms/Utils/Local.cpp:1032]: nvm i found a reduced testcase. Fix in #117778

**Comment by AreaZR** [comment]
Did we like backport this to 19.x? do we need to?

**Pull request merged**
